### ISSUE TITLE
Resolve `stress.py` errors to confirm SQLite use is fixed

### ISSUE THREAD

**Issue by jfalkner** [PacificBiosciences/smrtflow#161]
CC @skinner 

Followup work to #156. It appears SQLite use now works as expected on the server-side. No locking and the profiling of DB use shows all queries succeeding -- no errors. 

Yet `stress.py` still has issues. The work here is to sort out what is going on with the test. Hopefully this is  mostly a fix on the test ENV or in the test itself.

The setup here is to have one terminal with the server running.

```
# activate the virtualenv
source ENV/bin/activate

# clear out old DB data
rm -fr smrt-server-analysis/db

# run the server
make start-smrt-server-analysis > debug_db.txt
```

And one running the stress test.
```
# activate the virtualenv
source ENV/bin/activate

# setup needed data
make dataclean
make test-data/smrtserver-testdata

# run the stress test
(ENV)$ ./stress.py 2> stress.err > stress.out
```

Here are example logs of the issues I'm seeing.

[stress_err.txt](https://github.com/PacificBiosciences/smrtflow/files/314583/stress_err.txt)
[stress_out.txt](https://github.com/PacificBiosciences/smrtflow/files/314584/stress_out.txt)

**Comment by jfalkner** [comment]
I reran everything to get the latest logs at default logging level. 

`make start-smrt-server-analysis 2> server_err.txt > server_out.txt`

[server_err.txt](https://github.com/PacificBiosciences/smrtflow/files/314651/server_err.txt)
[server_out.txt](https://github.com/PacificBiosciences/smrtflow/files/314652/server_out.txt)

`./stress.py 2> stress.err > stress.out`

[stress_err.txt](https://github.com/PacificBiosciences/smrtflow/files/314653/stress_err.txt)
[stress_out.txt](https://github.com/PacificBiosciences/smrtflow/files/314654/stress_out.txt)

**Comment by jfalkner** [comment]
I've added some tooling in [`http_debugging`](https://github.com/PacificBiosciences/smrtflow/tree/http_debugging) to clarify what code is exercised on the server-side. Along with testing `stress.py` I'm going to see if removing the actor layer helps with this logic. I think the tooling can be generalized to something that logs/debugs all HTTP routes.

Based on the existing `server_out.txt` above the set of URLs can be pulled with:

```
# awk that shows all URL endpoints based on logging from our `RoutedHttpService` that is used by `ServiceActor`
awk '/http/{ print $0 }' server_out.txt
```

And after walking through them one-by-one, this `awk` confirms I've tooled them all.

```
# awk that filters out all tooled URL endpoints
awk '/http/ && !/job-manager\/jobs\/import-dataset/ && !/datasets\/references/ && !/secondary-analysis\/datasets\/.*-.*-.*-.*-.*/ && !/\/status/ && !/\/runs/ && !/job-manager\/jobs\/[0-9]+/{ print $0 }' server_out.txt
# ... no lines matched
```

Here ends up being a tally of usage. All of these start with `http://localhost:8070`.

```
# tally usage
awk 'match($0, /[GET|POST]+ :.*$/ )  { print substr($0, RSTART, RLENGTH) }' server_out.txt | sort -r | uniq -c

- POST
  - 10 : /secondary-analysis/job-manager/jobs/import-dataset : com.pacbio.secondary.smrtlink.services.jobtypes.ImportDataSetServiceType$$anonfun$3$$anonfun$apply$4$$anonfun$apply$5$$anonfun$apply$6.apply(ImportDataSetServiceType.scala:77)
  - 5 /smrt-link/runs : com.pacbio.secondary.smrtlink.services.RunService$$anonfun$2$$anonfun$apply$3$$anonfun$apply$4.apply(RunService.scala:51)
- GET
  - 15 /status : com.pacbio.common.services.StatusService$$anonfun$1$$anonfun$apply$1.apply(StatusService.scala:39)
  - 9 /secondary-analysis/job-manager/jobs/$ID : com.pacbio.secondary.smrtlink.services.JobService$$anonfun$sharedJobRoutes$2$$anonfun$apply$5$$anonfun$apply$6.apply(JobService.scala:117)
    - 3 /secondary-analysis/job-manager/jobs/36
    - 3 /secondary-analysis/job-manager/jobs/35
    - 3 /secondary-analysis/job-manager/jobs/34
    - 3 /secondary-analysis/job-manager/jobs/33
    - 3 /secondary-analysis/job-manager/jobs/32
    - 1 /secondary-analysis/job-manager/jobs/1
  - 11 /secondary-analysis/datasets/$UUID : com.pacbio.secondary.smrtlink.services.DataSetService$$anonfun$7$$anonfun$apply$9$$anonfun$apply$10.apply(DataSetService.scala:184)
    - 1 /secondary-analysis/datasets/f15bcd18-6c88-babf-a3c7-a9abb6f5a86f
    - 1 /secondary-analysis/datasets/a6254b4b-1a6a-d4d4-a217-10acaad9b77e
    - 1 /secondary-analysis/datasets/a25ae22f-2fa3-1f90-ed55-d76c851ceb59
    - 1 /secondary-analysis/datasets/8b4db227-5a21-4658-8c1a-cf0f190d38e9
    - 1 /secondary-analysis/datasets/63e85316-8ccb-039b-299d-fd7b4f75c3f0
    - 1 /secondary-analysis/datasets/5c8564bc-ac20-2070-3542-f6803bf42305
    - 1 /secondary-analysis/datasets/3513f44e-7cbe-29c4-ae78-09da67a92b79
    - 1 /secondary-analysis/datasets/23ade5fb-cdf0-9882-7296-0f6b4e959a02
    - 1 /secondary-analysis/datasets/17273d39-ebcf-1610-3a38-ce3e74ae4aa6
    - 1 /secondary-analysis/datasets/1382620c-92f9-57b9-373c-b60de2755512
    - 1 /secondary-analysis/datasets/111c1d1d-0fd1-aba2-6052-961b98d23a3b
  - 1 /secondary-analysis/datasets/references : com.pacbio.secondary.smrtlink.services.DataSetService$$anonfun$datasetRoutes$1$$anonfun$apply$2.apply(DataSetService.scala:100)

```

**Comment by jfalkner** [comment]
Error summary. Looks like almost all are timeouts.

35 instances of 
>Non-fatal exception: 500 - Internal Server Error - Ask timed out on [Actor[akka://smrtlink-analysis-server/user/$g#433770981]] after [10000 ms]

```
cat server_out.txt | grep Error | awk 'match($0, /akka[^\]]*/) { print substr($0, RSTART, RLENGTH) }' | sort | uniq -c | sort -r
  15 akka://smrtlink-analysis-server/user/$g#433770981
  15 akka://smrtlink-analysis-server/user/$f#-548514890
   5 akka://smrtlink-analysis-server/user/$h#1553688001
```

**Comment by jfalkner** [comment]
@skinner we were looking at timeouts. I just noticed that some of the runtimes for imports are 10 sec or more. e.g. `run_time=15.3`

> snapshot/stress_err.txt:120:JobExeError: Dataset d4c61804-f520-0728-8fe8-28e6a2abc7b7 was imported but could not be retrieved; this may indicate XML schema errors.
snapshot/stress_err.txt-121-ExtCmdResult(exit_code=2, cmd='pbservice import-dataset --host=localhost --port=8070 /Users/jfalkner/tokeep/git/PacificBiosciences/smrtflow/test-output/dataset-e5cef815-3dce-4ba1-b03c-c63e24620542/referenceset.xml', run_time=15.311681032180786)

**Comment by natechols** [comment]
That just reminded me that a few ReferenceSet XMLs are huge - up to several hundred MB - because they have thousands of contigs (short read "assemblies") and they all have XML tags.  I think the one used in stress.py only has 200 contigs; we should probably test the extreme case somewhere.

**Comment by jfalkner** [comment]
@natechols can you note a few that would be good to test? We should get a ticket up to roll those in to the stress test once it seems to work as-is.

**Comment by natechols** [comment]
I was a little off on my estimate of file size (it's the Python memory consumption that's hundreds of MB), but if you just list references by number of records in smrtlink-beta, you'll find a few that have tens of thousands of contigs.

**Comment by natechols** [comment]
PS. We should just generate this on the fly for testing, it's not worth putting the XML in git.

**Comment by jfalkner** [comment]
Here are some numbers from a `--nprocess 1` run of `stress.py`. All take over 10 secs

```
# run the server with INFO level logging
make start-smrt-server-analysis > server_out.txt 2> server_err.txt

# launch the stress test
$ ./stress.py --nprocesses 1 2> stress_err.txt > stress_out.txt

# tally all runtimes -- over 10 secs may cause Akka timeouts
$ grep -rn "run_time" stress_*.txt | awk 'match($0, /run_time=.{8}/) { print substr($0, RSTART, RLENGTH) }' | sort | uniq -c | sort -r
   4 run_time=23.12457
   4 run_time=23.12289
   4 run_time=23.12206
   4 run_time=23.12203
   4 run_time=23.11945
   4 run_time=23.11916
   4 run_time=23.11397
   4 run_time=17.12034
   4 run_time=12.01892
   4 run_time=12.01836
   4 run_time=12.01802
   4 run_time=12.01744
   4 run_time=12.01703
   4 run_time=12.01674
   4 run_time=12.01646
   4 run_time=12.01632
   4 run_time=12.01616
   4 run_time=12.01592
   4 run_time=12.01376
   2 run_time=10.03748
   2 run_time=10.03485
   2 run_time=10.03378
   2 run_time=10.03269
```

**Comment by skinner** [comment]
My understanding is that the imports run as jobs; the jobs are meant to be long-running and there's not supposed to be a live akka `ask` that spans the entire execution of the job.  So a long-running job should not result in an 'ask timed out'; there may be bugs, but jobs lasting > 10sec aren't *in themselves* evidence of a bug.

**Comment by skinner** [comment]
```
cat server_out.txt | grep Error | awk 'match($0, /akka[^\]]*/) { print substr($0, RSTART, RLENGTH) }' | sort | uniq -c | sort -r
  15 akka://smrtlink-analysis-server/user/$g#433770981
  15 akka://smrtlink-analysis-server/user/$f#-548514890
   5 akka://smrtlink-analysis-server/user/$h#1553688001
```

Logging the request on 500 errors (like [this](https://github.com/PacificBiosciences/smrtflow/pull/144/files#diff-cfe38c2461eb0fc94495b0c89b95e36bR153)) would provide some context about specifically what route is doing the `ask`.

Also, at one point I started looking at making our actor names more descriptive, but I got interrupted and dropped it.  Maybe it would be good though.

**Comment by jfalkner** [comment]
Still tracing the culprits of the timeouts. Recycling the same `awk` use from above.

```
$ cat server_out.txt | grep Error | awk 'match($0, /akka[^\]]*/) { print substr($0, RSTART, RLENGTH) }' | sort | uniq -c | sort -r
  20 akka://smrtlink-analysis-server/user/JobsDaoActor#336561178
   3 akka://smrtlink-analysis-server/user/RunServiceActor#107202521
   1 akka.actor.default-dispatcher-2
```

And all the errors are timeouts.

```
[info] akka.pattern.AskTimeoutException: Ask timed out on [Actor[akka://smrtlink-analysis-server/user/JobsDaoActor#336561178]] after [10000 ms]
```

**Comment by jfalkner** [comment]
I think `stress.py --nprocesses 1` is passing for me locally now with the latest master.

```
./stress.py --nprocesses 1 2> stress_err.txt > stress_out.txt 
...
Status Services have been up for 3 minutes and 29.066 seconds.
----------
Dataset Summary
----------
SubreadSets 5
HdfSubreadSets 0
ReferenceSets 6
AlignmentSets 0
----------
All Job types Summary
----------
import-dataset Jobs 11
State SUCCESSFUL 11
----------
merge-datasets Jobs 0
----------
convert-fasta-reference Jobs 0
----------
pbsmrtpipe Jobs 4
State SUCCESSFUL 4
----------
```

And no errors in `stress_out.txt` nor output in `stress_err.txt`

```
$ cat stress_out.txt | grep ERROR
$ cat stress_out.txt | grep Error
$ cat stress_out.txt | grep imeout
$ cat stress_err.txt 
```

The only errors in the server logs appear to be `IOError: Pipeline Summary Stats (sts.xml) not found or missing key distributions`

**Comment by jfalkner** [comment]
I'm seeing timeout errors in the following form when running `stress.py` and I can't seem to figure out how to increase the timeout value. I've tried explicitly setting `timeout` in the actors. Anyone have advice?

```
Error Futures timed out after [10 seconds]
```

Full message below.

>[info] DeadLetter(Failure(java.lang.Exception: Failed to import datastore Failed to add datastore file file DataStoreJobFile(2e828b18-fab6-40fb-a0dd-8258d6256273,DataStoreFile(d36a99b6-7578-03e6-ea94-238367a6a21d,pbscala::import_dataset,PacBio.DataSet.SubreadSet,4375,2016-06-21T13:59:59.449-04:00,2016-06-21T13:59:59.449-04:00,/Users/jfalkner/tokeep/git/PacificBiosciences/smrtflow/test-output/dataset-9e487d43-0a88-41d3-9845-51736fd99b54/subreads.xml,false,New-dataset-9e487d43-0a88-41d3-9845-51736fd99b54,Imported DataSet on 2016-06-21T13:59:59.449-04:00)) Error Futures timed out after [10 seconds]),Actor[akka://smrtlink-analysis-server/user/JobsDaoActor#1760265249],Actor[akka://smrtlink-analysis-server/deadLetters])

**Comment by jfalkner** [comment]
Ugh, implicits. I think I sorted it out.

**Comment by skinner** [comment]
If you sorted it out, cool, but in case it helps:

Asks happen in the actors and also in the route handlers (under the "services" directories in smrt-server-link and smrt-server-analysis).  A lot of the timeouts in the route handlers are inherited (with the intent of making it easier to change them by having them in fewer places, and under the assumption that it's reasonable to have the same timeouts for all the route handlers).

I just submitted a change to the S3.1 branch in perforce to try and reduce the error rate there; it includes some timeout bumps that I want to propagate to git/smrtflow:

http://swarm/changes/182032

There are some `implicit val timeout` values that the `ask` calls use, and one `Await.result` timeout.  IIRC I've seen the `Futures timed out` message from `Await.result` timeouts (the timeouts from `ask` calls result in an `ask timed out` message).  I don't know if/where there's an `Await.result` with a 10 second timeout though.

**Comment by jfalkner** [comment]
@skinner Thanks, I ended up walking through all the timeouts and setting them to unique values such as 10 seconds, 11 seconds, ... That clarified where the timeouts I needed to alter were set. It still isn't clear to me how to do this more easily in the future. It seemed that `implicit timeout` was changed in so many different spots that doing the above trick was the only helpful way to find the right one. For example, different methods on `JobsDaoActor.scala` seemed to use different timeouts -- despite me trying to explicitly set the timeout on that class.
